subsys/bluetooth/host/hci_core.c

```diff
@@ -5336,10 +5337,11 @@ static int le_adv_update(const struct bt_data *ad, size_t ad_len,
 		}
 
 		name = bt_get_name();
+		data = (struct bt_data)BT_DATA(
+			BT_DATA_NAME_COMPLETE,
+			name, strlen(name));
 
-		d[1].data = (&(struct bt_data)BT_DATA(
-					BT_DATA_NAME_COMPLETE,
-					name, strlen(name)));
+		d[1].data = &data;
```

Comment:
All this seems to be doing is changing from unnamed variable to use a variable, perhaps we are missing const so it is not destroyed when it is out of scope? Anyway I fine with the change, Im just trying to figure out why it would be a problem with some compilers.

Reply:
The original creates a pointer to a compiler-generated temporary that is destroyed when the scope is exited.  The pointer is stored in a structure defined in an enclosing scope and is invalid by the point it's used.

The fix holds the structure in a variable with the same lifetime as the pointer.

I don't see how it could ever have worked.

Reply:
Usually the compiler would attach the lifetime to the outer variable, there are plenty of example of that in the tree btw, perhaps it is the const modifier that does the trick, anyway if this is undefined behavior Im fine with the fix.

Reply:
No, const would have no effect on this.

The C11 final draft [n1570](http://www.open-std.org/jtc1/sc22/wg14/www/docs/n1570.pdf) in §6.5.2.5 defines compound literals and specifies that uses within a function have "automatic storage duration associated with the enclosing block".

The use at
https://github.com/zephyrproject-rtos/zephyr/blob/504dffa38c7933954877baeb9307c6f2fa0198d6/subsys/bluetooth/host/gatt.c#L2230 and similar inline argument construction is fine as long as the subsequent call to `params->func()` doesn't retain the pointer.  

But the use here, creating the pointer in an enclosed block and storing it in a pointer lvalue held in an enclosing block for use in that outer block, is simply wrong.

If you are aware of other uses that do that with `const` assuming that it's allowed, please correct them.

Reply:
> Usually the compiler would attach the lifetime to the outer variable, there are plenty of example of that in the tree btw, perhaps it is the const modifier that does the trick, anyway if this is undefined behavior Im fine with the fix.

> If you are aware of other uses that do that with const assuming that it's allowed, please correct them.

I tried to find other instances of this. There are a few i gatt.c, but all of them seems to have the correct scope in usage, so that should be fine.